### PR TITLE
Differential (Additive/Removal) patching when switching between design / debug

### LINKED ISSUE #2971: Updated requests rolled back to default after switching to DESIGN mode and then back to DEBUG mode
I have imported the design from the openAPI3.0 file.

I needed a few more requests and tried adding and reordering the requests in the debug mode. but every time I switch back to design mode, my changes are lost and the view resets to the default imported APIs

Here, I added the Auth folder and added a new request `Some New Request`, and ordered the `Login` API inside the `Auth` folder, and renamed it as `Reordered Login inside Auth Folder`.

image

When I switched to the design mode and returned back to the debug mode again, all the changes except newer requests are lost and the default environment variables are overridden.
image

It even clears the form data 

### before
image

### after
image

### Steps To Reproduce

1. Create a blank project
2. In the design tab, add OpenAPI 3.0 design
3. switch to the `DEBUG` tab
4. Reorder the requests (create a folder and move the request inside it, rename the request, etc.)
5. switch back to the `DESIGN` tab
6. again switch back to the `DEBUG` tab

### PR DESCRIPTION
**Pull request by apisaint**
😩 Finally got tired of dealing with it so I tried fixing it.

This patch adds differential patching for imports that occur during the switching between design and debug tabs inside of Designer. As reported through #2971, and others, this patch favors existing data over imported data, except urls (see below), values that exist already on the document will remain unchanged, only new values (including array based values) will be added / removed.

This also includes a bypass feature for urls, currently these options are not exposed through the interface but could be. This feature is an object to allow for future properties for preference-based patching.

This also allows endpoints to stay in position when moved or grouped in folders.

👇 Included:

- Adds `options.enableDiffBasedPatching` and `options.enableDiffDeep` to `importRaw` for backwards compat.
- Adds `options.bypassDiffProps.url` for url bypassing and use an object for future items.
- Adds `diffPatchObj` based on differential patching for objects (works with arrays as well).

💭 Future ideas:

- `hasBeenModifiedByUser` property map object to allow changing properties that haven't been touched by the user with options.
- Notice icon when clicked displays dialogue with changes that a user can choose to apply. Underlying diff method can be augmented to create such patch objects or use a lib.

⚠️ Notes:

Basically this makes changes additive, which is better than the current scenario especially when paired with #2979, what does that imply? See below.

- If it exists it won’t change unless a bypass option is created.
- When the `url` changes for a Request, when the bypass is enabled (true by default here) it will update the url in the url bar but it does not change the `name`, so the sidebar will not auto-update. This can be resolved in a future or with a modified check for the name using the future idea explained above.
- When items / values inside of the Headers change it does not auto-update, this would require either a bypass option, or rely on the future idea.

Related #3038
Closes #2971, #2882, #2442

### CODE REVIEW

**Comment by reynolek** [comment]
We have this on our list of issues to tackle, thanks for taking the time to look into it! Going to add @nijikokun to review it from a product perspective along with @develohpanda and @dimitropoulos for code.

File: packages/insomnia-app/app/common/import.js [hunk 1]
```diff
@@ -240,7 +241,19 @@ export async function importRaw(
     const existingDoc = await db.get(model.type, resource._id);
     let newDoc: BaseModel;
     if (existingDoc) {
-      newDoc = await db.docUpdate(existingDoc, resource);
+      let updateDoc = resource;
+
+      // Do differential patching when enabled
+      if (options.enableDiffBasedPatching) {
```
Comment (nijikokun) [inline packages/insomnia-app/app/common/import.js]: `options` isn't checked for existence so this will fail with other import methods.

**Review by nijikokun — CHANGES_REQUESTED** [review]

**Comment by nijikokun** [comment]
Oh wow. Taking a look now  👀 

> * backwards compat.

Beautiful 🙌 

> This also allows endpoints to stay in position when moved or grouped in folders.

Absolutely amazing 🎉 

> * it will update the url in the url bar but it does not change the `name`, so the sidebar will not auto-update. This can be resolved in a future or with a modified check for the name using the future idea explained above.

While we do have `modified` you're right we don't have something that tells us whether it's been modified by the user or through creation, at least to my knowledge, so I agree this can be a future improvement. 👍 

> * Notice icon when clicked displays dialogue with changes that a user can choose to apply. Underlying diff method can be augmented to create such patch objects or use a lib.

Love this idea 💯, maybe this could also be expanded to sync at some point 🤔

> * `hasBeenModifiedByUser` property map object to allow changing properties that haven't been touched by the user with options.

Makes sense, track properties that change, and update the ones that haven't, seems like a good future improvement.  💯 

> * Adds `options.bypassDiffProps.url` for url bypassing and use an object for future items.

👍 

---

Amazing, love the breakdown. Found an issue in the code, but other than that it's definitely an improvement over the current situation and I'll take it.

File: packages/insomnia-app/app/common/import.js (hunk 1 above)
Comment (apisaint) [reply packages/insomnia-app/app/common/import.js]: Good catch, instead of worrying about existence decided to default the argument to an object and let type checking do the worrying about input for us.

**Review by nijikokun — APPROVED** [review]

**Review by dimitropoulos — APPROVED** [review]

**Comment by dimitropoulos** [comment]
(working on making sure I correctly resolve the merge conflict of the prior commit on master then this is next up to be merged)
